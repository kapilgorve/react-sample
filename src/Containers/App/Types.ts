export interface IStatement {
    date: string;
    monthlyInterest: number;
    monthlyPrincipal: number;
    balance: number;
}