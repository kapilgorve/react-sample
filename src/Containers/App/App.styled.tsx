import styled from 'styled-components';

export const Button = styled.button`
  display: inline-block;
  color: palevioletred;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid palevioletred;
  border-radius: 3px;
  display: block;
`;

export const BodyContainer = styled.div`
  background: #e4edf2;
  padding 15px;
`;

export const Container = styled.div`
  padding 15px;
`;

export const Form = styled.form`
  width: 50%;
  max-width: 600px;
  background: white;
  border-radius: 12px;
  margin: auto;
  padding: 2rem;
  margin-top: -20px;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 3px 6px 0px;
`;

export const FormGroup = styled.div`
  display: flex;
  align-items: center;
  margin: 24px 0;
`;

export const FormInput = styled.input`
  width: 100%;
  font-size: 24px;
  color: #758698;
  padding: 8px;
  border: solid 1px #e4edf2;
  border-radius: 8px;
`;

export const Label = styled.label`
  font-size: 24px;
  color: #758698;
  min-width: 40%;
`;

export const EmiTitle = styled.h3`
  color: #251f20;
  font-size: 40px;
  text-align: center;
  margin: 3rem;
  font-weight: lighter;
`;

export const Card = styled.div`
  background: white;
  border-radius: 12px;
  margin: 2rem;
  padding: 2rem;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 3px 6px 0px;
`;

export const CardWrap = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
`;

export const TermName = styled.h3`
  font-size: 24px;
  font-weight: lighter;
  color: #758698;
  margin: 2rem;
`;

export const TermValue = styled.p`
  font-size: 24px;
  font-weight: lighter;
  color: #758698;
  margin: 2rem;
`;

export const Wrap = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const CardTitle = styled.h3`
  font-size: 30px;
  font-weight: lighter;
  color: #758698;
  text-align: center;
  margin: 0 0 30px;
`;

export const Table = styled.table`
  width: 100%;
  margin-bottom: 1rem;
  color: #212529;
  border-collapse: collapse;
`;

export const Td = styled.td`
  text-align: left;
  border: 1px solid #ddd;
  padding: 12px;
`;

export const Th = styled.th`
  text-align: left;
  border: 1px solid #ddd;
  padding: 12px;
  border-bottom: 2px solid #dee2e6;
`;