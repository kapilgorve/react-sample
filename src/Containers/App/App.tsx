import React, { useState, useEffect } from 'react';
import { calculateEmi, calculateMonthly } from '../../util/calculator';
import {
  BodyContainer,
  Form,
  FormInput,
  Label,
  FormGroup,
  EmiTitle,
  Card,
  CardWrap,
  TermName,
  TermValue,
  Wrap,
  CardTitle,
  Table,
  Th,
  Td,
} from './App.styled';
import PieChart from '../../Components/PieChart';
import Header from '../../Components/Header';
import './App.css';
import { IStatement } from './Types';

const App = () => {
  const [form, setForm] = useState({
    principal: 10000,
    rate: 10,
    tenure: 1,
  });
  const [loanTerms, setLoanTerms] = useState({
    emi: 0,
    interest: 0,
    totalPayable: 0,
  });

  const [statements, setStatements] = useState([] as IStatement[]);

  useEffect(() => {
    const { principal, rate, tenure } = form;
    let terms = calculateEmi(principal, rate, tenure);
    setLoanTerms(terms);
    let monthly = calculateMonthly(principal, rate, tenure, terms.emi);
    setStatements(monthly);
  }, [form]);

  const updateField = (e) => {
    const { name, value } = e.target;
    const parsed = parseInt(value);
    if (parsed) {
      setForm({
        ...form,
        [name]: parsed,
      });
    }
  };

  const renderForm = () => {
    return (
      <Form>
        <FormGroup>
          <Label>Loan Amount</Label>
          <FormInput name="principal" value={form.principal} onChange={updateField} data-testid="principalInput" />
        </FormGroup>
        <FormGroup>
          <Label>Interest Rate (%)</Label>
          <FormInput name="rate" value={form.rate} onChange={updateField} data-testid="rateInput" />
        </FormGroup>
        <FormGroup>
          <Label>Tenure (Years)</Label>
          <FormInput name="tenure" value={form.tenure} onChange={updateField} data-testid="tenureInput" />
        </FormGroup>
      </Form>
    );
  };

  const renderLoanDetails = () => {
    return (
      <Card>
        <Wrap>
          <TermName>Loan EMI</TermName>
          <TermValue>{Math.round(loanTerms.emi)}</TermValue>
        </Wrap>
        <Wrap>
          <TermName>Total Interest</TermName>
          <TermValue data-testid="interest">{loanTerms.interest}</TermValue>
        </Wrap>
        <Wrap>
          <TermName>Prinicipal</TermName>
          <TermValue data-testid="principal">{form.principal}</TermValue>
        </Wrap>
        <Wrap>
          <TermName>Total payable</TermName>
          <TermValue data-testid="payable">{loanTerms.totalPayable}</TermValue>
        </Wrap>
      </Card>
    );
  };

  const renderTable = () => {
    return (
      <Card>
        <Table>
          <thead>
            <tr>
              <Th>Month/Year</Th>
              <Th>Prinicipal</Th>
              <Th>Interest</Th>
              <Th>Balance</Th>
            </tr>
          </thead>
          <tbody>
            {statements &&
              statements.map((monthly, index) => {
                return (
                  <tr key={index}>
                    <Td>{monthly.date}</Td>
                    <Td>{Math.round(monthly.monthlyPrincipal)}</Td>
                    <Td>{Math.round(monthly.monthlyInterest)}</Td>
                    <Td>{Math.round(monthly.balance)}</Td>
                  </tr>
                );
              })}
          </tbody>
        </Table>
      </Card>
    );
  };

  return (
    <>
      <Header />
      <BodyContainer>
        {renderForm()}
        <EmiTitle>
          Your Monthly Loan EMI: <span data-testid="emititle">{Math.round(loanTerms.emi)}</span>
        </EmiTitle>
        <CardWrap>
          {renderLoanDetails()}
          <Card>
            <CardTitle> Pie Chart</CardTitle>
            <PieChart interest={loanTerms.interest} totalPayable={loanTerms.totalPayable} />
          </Card>
        </CardWrap>
        {renderTable()}
      </BodyContainer>
    </>
  );
};

export default App;
