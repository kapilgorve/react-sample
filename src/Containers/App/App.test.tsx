import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';

import App from './App';

afterEach(cleanup);

describe('<App/> Component', () => {
  it('renders <App/> ', () => {
      const {asFragment} =  render(<App/>)
      expect(asFragment()).toMatchSnapshot();
  });

  it('Enter principal', async () => {
    const { getByTestId } = render(<App />);
    const principalInputElement = getByTestId('principalInput') as HTMLInputElement;
    fireEvent.change(principalInputElement, { target: { value: '20000' } });
    expect(getByTestId('emititle')).toHaveTextContent('1758');
  });

  it('Enter rate', async () => {
    const { getByTestId } = render(<App />);
    const rateInputElement = getByTestId('rateInput') as HTMLInputElement;
    fireEvent.change(rateInputElement, { target: { value: '12' } });
    expect(getByTestId('emititle')).toHaveTextContent('888');
  });

  it('Enter tenure', async () => {
    const { getByTestId } = render(<App />);
    const tenureInputElement = getByTestId('tenureInput') as HTMLInputElement;
    fireEvent.change(tenureInputElement, { target: { value: '2' } });
    expect(getByTestId('emititle')).toHaveTextContent('461');
  });

  it('Enter principal, rate , tenure', async () => {
    const { getByTestId } = render(<App />);
    const principalInputElement = getByTestId('principalInput') as HTMLInputElement;
    const rateInputElement = getByTestId('rateInput') as HTMLInputElement;
    const tenureInputElement = getByTestId('tenureInput') as HTMLInputElement;
    fireEvent.change(principalInputElement, { target: { value: '30000' } });
    fireEvent.change(rateInputElement, { target: { value: '10' } });
    fireEvent.change(tenureInputElement, { target: { value: '3' } });
    expect(getByTestId('emititle')).toHaveTextContent('968');
  });

  it('Check Interest', async () => {
    const { getByTestId } = render(<App />);
    const principalInputElement = getByTestId('principalInput') as HTMLInputElement;
    fireEvent.change(principalInputElement, { target: { value: '20000' } });
    expect(getByTestId('interest')).toHaveTextContent('1100');
  });

  it('Check payable', async () => {
    const { getByTestId } = render(<App />);
    const principalInputElement = getByTestId('principalInput') as HTMLInputElement;
    fireEvent.change(principalInputElement, { target: { value: '20000' } });
    expect(getByTestId('payable')).toHaveTextContent('21100');
  });

});
