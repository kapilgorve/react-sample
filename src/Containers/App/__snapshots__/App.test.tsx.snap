// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> Component renders <App/>  1`] = `
<DocumentFragment>
  <div
    class="sc-fzqNJr eTiVkk"
  >
    <h2
      class="sc-fzoyAV irMlCl"
    >
      EMI Calculator
    </h2>
    <h4
      class="sc-fzoLag cwxDVD"
    >
      Calculate your Loan EMI & Total Interest Due
    </h4>
  </div>
  <div
    class="sc-AxirZ epPGuK"
  >
    <form
      class="sc-AxhCb clIZyw"
    >
      <div
        class="sc-AxhUy bAEQme"
      >
        <label
          class="sc-AxheI dPheQk"
        >
          Loan Amount
        </label>
        <input
          class="sc-AxgMl ldIRBI"
          data-testid="principalInput"
          name="principal"
          value="10000"
        />
      </div>
      <div
        class="sc-AxhUy bAEQme"
      >
        <label
          class="sc-AxheI dPheQk"
        >
          Interest Rate (%)
        </label>
        <input
          class="sc-AxgMl ldIRBI"
          data-testid="rateInput"
          name="rate"
          value="10"
        />
      </div>
      <div
        class="sc-AxhUy bAEQme"
      >
        <label
          class="sc-AxheI dPheQk"
        >
          Tenure (Years)
        </label>
        <input
          class="sc-AxgMl ldIRBI"
          data-testid="tenureInput"
          name="tenure"
          value="1"
        />
      </div>
    </form>
    <h3
      class="sc-Axmtr eBLoLY"
    >
      Your Monthly Loan EMI: 
      <span
        data-testid="emititle"
      >
        879
      </span>
    </h3>
    <div
      class="sc-fzozJi fpzqXb"
    >
      <div
        class="sc-AxmLO hJIyVP"
      >
        <div
          class="sc-fzplWN CdBNd"
        >
          <h3
            class="sc-fzoLsD lloRMf"
          >
            Loan EMI
          </h3>
          <p
            class="sc-fzpans cecatz"
          >
            879
          </p>
        </div>
        <div
          class="sc-fzplWN CdBNd"
        >
          <h3
            class="sc-fzoLsD lloRMf"
          >
            Total Interest
          </h3>
          <p
            class="sc-fzpans cecatz"
            data-testid="interest"
          >
            550
          </p>
        </div>
        <div
          class="sc-fzplWN CdBNd"
        >
          <h3
            class="sc-fzoLsD lloRMf"
          >
            Prinicipal
          </h3>
          <p
            class="sc-fzpans cecatz"
            data-testid="principal"
          >
            10000
          </p>
        </div>
        <div
          class="sc-fzplWN CdBNd"
        >
          <h3
            class="sc-fzoLsD lloRMf"
          >
            Total payable
          </h3>
          <p
            class="sc-fzpans cecatz"
            data-testid="payable"
          >
            10550
          </p>
        </div>
      </div>
      <div
        class="sc-AxmLO hJIyVP"
      >
        <h3
          class="sc-fznyAO bnaQsZ"
        >
           Pie Chart
        </h3>
        <div
          class="sc-fzqBZW laeCLi"
        >
          <div>
            <canvas
              class="chartjs-render-monitor"
              height="0"
              style="display: block;"
              width="0"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-AxmLO hJIyVP"
    >
      <table
        class="sc-fznKkj eUDXGF"
      >
        <thead>
          <tr>
            <th
              class="sc-fzokOt kDWarP"
            >
              Month/Year
            </th>
            <th
              class="sc-fzokOt kDWarP"
            >
              Prinicipal
            </th>
            <th
              class="sc-fzokOt kDWarP"
            >
              Interest
            </th>
            <th
              class="sc-fzokOt kDWarP"
            >
              Balance
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Apr 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              796
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              83
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              9204
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              May 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              802
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              77
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              8402
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Jun 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              809
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              70
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              7593
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Jul 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              816
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              63
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              6777
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Aug 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              823
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              56
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              5954
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Sep 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              830
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              50
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              5124
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Oct 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              836
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              43
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              4288
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Nov 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              843
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              36
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              3445
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Dec 2020
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              850
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              29
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              2594
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Jan 2021
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              858
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              22
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              1737
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Feb 2021
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              865
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              14
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              872
            </td>
          </tr>
          <tr>
            <td
              class="sc-fznZeY bPECWU"
            >
              Mar 2021
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              872
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              7
            </td>
            <td
              class="sc-fznZeY bPECWU"
            >
              0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
